hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/utils/SignatureMapUtils.java

```diff
@@ -0,0 +1,84 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hedera.node.app.service.contract.impl.utils;
+
+import static java.util.Objects.requireNonNull;
+
+import com.hedera.hapi.node.base.SignatureMap;
+import com.hedera.hapi.node.base.SignaturePair;
+import com.hedera.hapi.node.base.SignaturePair.SignatureOneOfType;
+import com.hedera.node.app.spi.signatures.SignatureVerifier;
+import com.hedera.pbj.runtime.OneOf;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Some utility methods that are useful for processing system contracts.
+ */
+public class SignatureMapUtils {
+    private SignatureMapUtils() {
+        // Utility class
+        throw new UnsupportedOperationException("Utility Class");
+    }
+
+    /**
+     * When a signature map is passed to a system contract function and contains ECDSA signatures, before the signatures
+     * can be verified with a {@link SignatureVerifier}, they must be preprocessed in the following way:
+     * (1) If the v value is greater than 35, it must be checked to see if it matches the chain ID per EIP 155
+     * (2) Strip the v value from the public key as it is not needed for verification
+     * @param sigMap
+     * @return a new SignatureMap with the ECDSA signatures preprocessed
+     */
+    public static SignatureMap preprocessEcdsaSignatures(@NonNull final SignatureMap sigMap, final int chainId) {
+        final List<SignaturePair> newPairs = new ArrayList<>();
+        for (var spair : sigMap.sigPair()) {
+            if (spair.hasEcdsaSecp256k1()) {
+                final var ecSig = requireNonNull(spair.ecdsaSecp256k1());
+                if (ecSig.length() > 64) {
+                    if (!validChainId(ecSig.toByteArray(), chainId)) {
+                        throw new IllegalArgumentException("v value in ECDSA signature does not match chain ID");
+                    }
+                    spair = new SignaturePair(
+                            spair.pubKeyPrefix(), new OneOf<>(SignatureOneOfType.ECDSA_SECP256K1, ecSig.slice(0, 64)));
+                }
+            }
+            newPairs.add(spair);
+        }
+        return new SignatureMap(newPairs);
+    }
+
+    /**
+     * Check that the v value in an ECDSA signature matches the chain ID if it is greater than 35 per EIP 155
+     * @param ecSig
+     * @param chainId
+     * @return true if the v value matches the chain ID or is not relevant, false otherwise
+     */
+    public static boolean validChainId(final byte[] ecSig, final int chainId) {
+        int v = 0;
+        for (int i = 64; i < ecSig.length; i++) {
+            v <<= 8;
+            v |= (ecSig[i] & 0xFF);
```

Comment:
Always makes me happy to see compound assignment operators, especially rare ones like `<<=`.